Refactored interface

autumn-moon opened the pull request:
I was looking at the interface with the many parameters and I thought that it is superfluous hence the alternative taking the RenderOptions object. I think that is the more suitable approach. Basically I've set the other method deprecated and swapped the signatures so that the deprecated method now calls the one taking the InputStream and the RenderOptions.

I also wanted to refactor the render() method itself a bit. I tried to make it a little more readable and for the validity check I added BeanValidation (JSR-303).

I did some additional refactorings e.g. in naming.
Hope you like the PR.

Autumn-Moon committed:
Refactored interface

moodysalem commented on pom.xml:
```diff
@@ -120,6 +120,19 @@
+      <!-- Bean validation -->
+      <dependency>
+        <groupId>org.hibernate</groupId>
+        <artifactId>hibernate-validator</artifactId>
```
do we really need a hibernate module to do validation? i thought this added things like hibernate precommit and prepersist

autumn-moon replied:
This is the reference implementation for JSR-303 Bean Validation offered by hiberante project. Not to confuse with any of their JPA persistence stuff.
However you can integrate it into JPA. If you use JPA your Bean Validation is automatically checked on PrePersist and PreUpdate as far as I know.

moodysalem commented on wrapper/src/main/java/com/moodysalem/phantomjs/wrapper/PhantomJS.java:
```diff
@@ -88,15 +69,43 @@ public static InputStream render(final InputStream html, final RenderOptions opt
 
-        if (jsWait < 0 || jsInterval < 0 || (jsWait > 0 && jsInterval > jsWait) || (jsInterval == 0 && jsWait > 0)) {
-            throw new IllegalArgumentException("Invalid jsWait or jsInterval values provided");
+        if(!options.isValid()) {
```
i'd rather keep the package slim and explicitly do the validations than add a dependency and make this isValid method opaque-what's the advantage of annotation driven validation if we are only validating one class in this package?

autumn-moon replied:
I see your point. However I like the idea of the beans holding their own constraints instead of keeping the constraints elsewhere in the code. Especially stuff like the constraint between jsExecutionTimeout and jsInterval.

moodysalem replied:
explicitly throwing an exception also seems funny here if we are using hibernate validator-shouldn't it generate the exception?

still don't want to bring in hibernate validator-it doesn't give us much and could cause issues with other people's dependencies

moodysalem commented on wrapper/src/main/java/com/moodysalem/phantomjs/wrapper/PhantomJS.java:
```diff
@@ -121,16 +137,28 @@ public static InputStream render(final PhantomJSOptions options, final InputStre
-                new CommandLineArgument(wrapCommandLineArgumentName(PhantomJSConstants.JSINTERVAL_TEMPLATENAME), PhantomJSConstants.JSINTERVAL_TEMPLATENAME, jsInterval));
+        final Path renderPath = PhantomJSConstants.TEMP_RENDER_DIR.resolve(String.format(PhantomJSConstants.TARGET_PREFIX + "%s.%s", renderId, options.getRenderFormat().name().toLowerCase()));
+
+        List<CommandLineArgument> commandLineArguments = new ArrayList<>();
```
an `ArrayList` isn't as good here as a `LinkedList`, since we are doing adds and unsure about the total size.

moodysalem replied:
also, prefer to use the keyword `final` any time we do not intend to manipulate the variable after initialization

moodysalem replied:
> unsure about total size 
actually the # and position of arguments for the render script are fixed, so you could initialize your array list with the correct initial capacity, but i'm not sold on using a list and converting to an array over using varargs -
 the varargs style makes it more transparent that you're calling a phantomjs binary with some set of arguments in a particular order

autumn-moon replied:
I agree. I could change that to a fixed array instead of the List and conversion.
Regarding the final: I totally agree. I switched to intellij and I didn't activate the saveActions plugin. This comes for free in eclipse. ;)

moodysalem replied:
i think better than adding to an array list, you could just create a commandlineargument array and set indices in that array.

moodysalem replied:
that has the benefit of being explicit on which arguments go in which positions

moodysalem commented on wrapper/src/main/java/com/moodysalem/phantomjs/wrapper/beans/RenderOptions.java:
```diff
@@ -1,61 +1,92 @@
-        this.jsWait = jsWait;
-        this.jsInterval = jsInterval;
+
+    private static Validator validator = Validation.buildDefaultValidatorFactory().getValidator();
```
we definitely do not want this class to be mutable-reuse of options is extremely likely and should be encouraged, and creating new instances of this class that have the same set of options is going to be a waste in most use cases

autumn-moon replied:
I think either mutable or immutable is fine. I didn't see any benefit in mutable object having fluent API, so I changed it to a fitting mutable one. And I didn't like the awfully long parameter list in the constructor that was the reason I chose a mutable object over an immutable.

moodysalem replied:
we should keep it immutable and we can make the constructor with many arguments private-the fluent api is nice because it lets you define your immutable options object in a single line. this is a nice pattern to follow https://commons.apache.org/proper/commons-csv/apidocs/org/apache/commons/csv/CSVFormat.html

moodysalem commented on wrapper/src/test/java/com/moodysalem/java/TestPhantomJS.java:
```diff
@@ -38,20 +36,21 @@ public void testRunScript() throws IOException, InterruptedException {
-        assertTrue(PhantomJS.exec(TestPhantomJS.class.getResourceAsStream("test-script.js"), PhantomJSOptions.DEFAULT.withHelp(true)).getExitCode() == 0);
-        assertTrue(PhantomJS.exec(TestPhantomJS.class.getResourceAsStream("test-script.js"), PhantomJSOptions.DEFAULT.withVersion(true)).getExitCode() == 0);
+        assertTrue(PhantomJS.exec(TestPhantomJS.class.getResourceAsStream("test-script.js"), PhantomJsOptions.DEFAULT.withHelp(true)).getExitCode() == 0);
+        assertTrue(PhantomJS.exec(TestPhantomJS.class.getResourceAsStream("test-script.js"), PhantomJsOptions.DEFAULT.withVersion(true)).getExitCode() == 0);
```
because PhantomJsOptions is mutable in this PR, the meaning of this test has changed-now it calls help twice

autumn-moon replied:
You mean RenderOptions, right? I didn't change the PhantomJsOptions despite changing it's name to camel case.

moodysalem replied:
gotcha, haven't looked at the code in a while, assumed they were both mutable as of the PR

moodysalem commented:
I like making the old interface deprecated and making it call the better interface but bringing in the hibernate validator to save a couple of lines of very easy to understand code doesn't seem worth it IMO, and I think it's valuable for options objects to be immutable. 

I'd consider maybe a builder class separate from the options objects to remove the `with` functions from the options objects, but that's a marginal improvement (and some people might say it's not an improvement)

ghost commented:
Thank you for your time reviewing my PR. I will amend the fixed array. I'd like to hear your opinion on my comments above.

Autumn-Moon committed:
Amendmens from PR

+ Extended .gitignore
+ Removed bean validation
+ Made RenderOptions immutable again

Autumn-Moon committed:
Used a fixed array for command line arguments

ghost commented:
I think I got all your comments covered. Need to figure out why rendering is succesful for the test which expects rendering to fail due to timeout settings.

moodysalem commented on wrapper/src/main/java/com/moodysalem/phantomjs/wrapper/RenderOptionsException.java:
```diff
@@ -1,25 +1,11 @@
-    });
 
-    return messageBuider.toString();
+  public RenderOptionsException(String message) {
```
this seems no longer necessary

moodysalem commented on wrapper/src/main/java/com/moodysalem/phantomjs/wrapper/beans/RenderOptions.java:
```diff
@@ -1,248 +1,151 @@
-
-        return constraintViolations;
-    }
+  private PhantomJsOptions phantomJsOptions = PhantomJsOptions.DEFAULT.withDiskCache(true);
```
shouldn't these be final since the setters return new instances?

moodysalem replied:
you will have to initialize them only in the constructor to make these final
